### ISSUE TITLE
OmniSci backend crashes with non-string column labels

### ISSUE THREAD

**Issue by dchigarev** [modin-project/modin#3370]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux
- **Modin version** (`modin.__version__`): 57bcfc1
- **Python version**: 3.7.10
- **Code we can use to reproduce**:
```python
import modin.pandas as pd
import traceback

def synthetic_example(non_str_columns_df):
    any_index = pd.Index([1, 2])
    # `df.index = any_index` defaults to pandas, so to reproduce
    # the error it's required to call `modin_frame._set_index` directly:
    non_str_columns_df._query_compiler._modin_frame._set_index(any_index)

def natural_example(non_str_columns_df):
    # `query_compiler._agg` does `modin_frame._set_index` call
    non_str_columns_df.sum()

non_str_columns_df = pd.DataFrame([[1, 2, 3], [3, 4, 5]], columns=[0, 1, 2])

try:
    synthetic_example(non_str_columns_df)
except Exception:
    print(f"Synthetic example:\n{traceback.format_exc()}")

try:
    natural_example(non_str_columns_df)
except Exception:
    print(f"\nNatural example:\n{traceback.format_exc()}")

```

<details><summary>Output</summary>

```
Synthetic example:
Traceback (most recent call last):
  File "t2.py", line 45, in <module>
    synthetic_example(non_str_columns_df)
  File "t2.py", line 36, in synthetic_example
    non_str_columns_df._query_compiler._modin_frame._set_index(any_index)
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/engines/omnisci_on_ray/frame/data.py", line 1962, in _set_index
    index_at = index_at.rename_columns(index_names + list(self.columns))
  File "pyarrow/table.pxi", line 1903, in pyarrow.lib.Table.rename_columns
  File "stringsource", line 15, in string.from_py.__pyx_convert_string_from_py_std__in_string
TypeError: expected bytes, int found

Natural example:
Traceback (most recent call last):
  File "t2.py", line 50, in <module>
    natural_example(non_str_columns_df)
  File "t2.py", line 40, in natural_example
    non_str_columns_df.sum()
  File "/localdisk/dchigare/repos/modin_bp/modin/pandas/dataframe.py", line 2076, in sum
    **kwargs,
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/backends/omnisci/query_compiler.py", line 96, in method_wrapper
    return method(self, *args, **kwargs)
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/backends/omnisci/query_compiler.py", line 362, in sum
    return self._agg("sum", **kwargs)
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/backends/omnisci/query_compiler.py", line 96, in method_wrapper
    return method(self, *args, **kwargs)
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/backends/omnisci/query_compiler.py", line 416, in _agg
    pandas.Index.__new__(pandas.Index, data=["__reduced__"], dtype="O")
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/engines/omnisci_on_ray/frame/data.py", line 1962, in _set_index
    index_at = index_at.rename_columns(index_names + list(self.columns))
  File "pyarrow/table.pxi", line 1903, in pyarrow.lib.Table.rename_columns
  File "stringsource", line 15, in string.from_py.__pyx_convert_string_from_py_std__in_string
TypeError: expected bytes, int found
```

</details>

### Describe the problem
When working with non-string column labels it could happen that some backend logic would try to insert a column with a string name to the frame, this will result in producing the arrow Table with the string and non-string column names, which is not supported by pyarrow:
```python
>>> import pyarrow as pa
>>> at = pa.Table.from_pydict({"1": [1, 2], 1: [3, 4]})
TypeError: expected bytes, int found
```

**Comment by YarShev** [comment]
@dchigarev , we have respective restrictions on this ([link](https://modin.readthedocs.io/en/latest/flow/modin/experimental/engines/omnisci_on_ray/omnisci_engine.html#relational-engine-limitations)). What would you like to do as part of this issue?

**Comment by dchigarev** [comment]
@YarShev at least this instead of an error:
> "When the unsupported data type is detected or unsupported operations is requested we use the original pandas framework."

Also, the reproducers above don't violate the rule about the homogeneous data, the frame with the mixed column names is produced by the **backend** logic, so it's not users fault of getting this error. We should avoid producing conflicting frames somehow

**Comment by YarShev** [comment]
Closed by #4767.
